6/1 Stopping Point

samiklein011011 opened the pull request:
Hi Tyler! So I added a table in mySQL for comments and a table for tags. Each post now can display the tags and comments that correspond with it. I also made it so that you can filter posts by tags and filter by author of the post. I also created functions for a lot of the things I was doing. For example, I made a function that can print an entire blog post, given the blog post ID. Let me know what you think!

(also I really hope this is the right way to make a pull request now!)

Commit:
5/25

Converted Blog Posts into an array on PHP. good stopping point

Commit:
5/28

Tyler explained a better way to do a for loop. Committing because I want to look at what I had (for loop). About to build something better (foreach).

Commit:
5/29 basics of website completed

Thought this was a good stopping point. I think the basics of the website completed are mostly completed. Still can work on it to make it more aestectically pleasing though.

Commit:
committing before making new branch

5/30

Commit:
5/30, comments work!

woo! Very exciting!

Commit:
5/31 Tags almost done!!!

AHHHH!! I made a function for print post, then made it so you could filter by tags! Made a different kind of frame for viewing blog posts, in which multiple could come up. So that you can see all the "food" posts at once. But the tag line is still messed up

Commit:
6/1 Tags, Comments, and Filters completed

Now each post has comments (if there are them in SQL). You can now filter by tags or filter by author. I also created functions for a ton of stuff (like even a function to print an entire blog post! Woo!)

tilikang commented on 17thblog/blogByAuthor.php:
```diff
@@ -0,0 +1,21 @@
+
+// foreach post within this author, print it
+
+	$posts=getPostsWithThisAuthor($_REQUEST['authorOfPost']);
```
This is super minor and I won't comment on it elsewhere in this PR, but this shouldn't be indented. Normally you'd only indent code if it's inside  curly braces `{}` like an if statement, function, or loop.

Maybe you know this, but you can easily clean something like this by selecting the whole block of text and pressing tab to indent or shift+tab to un-indent.

tilikang commented on 17thblog/blogByAuthor.php:
```diff
@@ -0,0 +1,21 @@
+	$posts=getPostsWithThisAuthor($_REQUEST['authorOfPost']);
+
+	foreach($posts as $post){
+		$var = $post['blogPostID'];
```
This is a totally fine way to write this, but I just want to point out a one-line way to write this same thing:

```
echo printBlogPost($post['blogPostID']);
```

tilikang commented on 17thblog/blogByTagsFrame.php:
```diff
@@ -0,0 +1,21 @@
+
+// foreach post within this tag, print it
+
+	$tags=getTagsWithThisName($_REQUEST['tag']);
```
This looks good. If you're looking for ideas on how to make your code a bit more robust/professional, something to consider is adding error handling. For example, what if someone comes to this page without having a `tag` defined in the URL parameters. Maybe you could show an error, redirect to the tag index page, etc.

tilikang commented on 17thblog/blogByTagsFrame.php:
```diff
@@ -0,0 +1,21 @@
+
+	foreach($tags as $tag){
+		$var = $tag['blogPostID'];
+		$toPrint = printBlogPost($var);
```
This function makes it look like you're printing blog posts, not tags. If that's true, I'd suggest renaming stuff above. `getTagsWithThisName()` should probably be `getPostsWithThisTag()`. And `foreach($tags as $tag){` should probably be `foreach($posts as $posts){`. This doesn't change the functionality, but it makes the code more readable. As I was reading this, I thought we were looping through a list of tags, not a list of blog posts.

tilikang commented on 17thblog/blogByAuthor.php:
```diff
@@ -0,0 +1,21 @@
+$wantToWrite="Author: $_REQUEST[authorOfPost]";
+
+echoNavAndHead($wantToWrite);
+printPageName($wantToWrite);
```
Is there a situation where you call one of these functions, but not both? If so, this is a fine approach. But if you always call both of them, I'd suggest just combining them into one function. You could take a single parameter (`$title` maybe) and use it for the header title and h1 tag, or you could take two separate parameters just so you have the flexibility to make them different if you want.

tilikang replied:
also, a naming suggestion: I'd either use `echo` or `print` as the first word in these function names, but I wouldn't switch back and forth. You can choose the naming convention you prefer, but consistency is important.

samiklein011011 replied:
I call both functions together on almost every page except the home page, which is why I had them as separate functions. So I could either keep calling both separately, or create one function that combines them, and create a new function that spits out the nav bar just for the home page. I'm not sure which one is best though.

samiklein011011 replied:
Also will definitely keep the consistency with naming from now on!

tilikang replied:
I'd either make a single function that then calls these two separate functions (so you don't need to call them both on most pages) or make a single file that takes an optional parameter for whether or not it should print the page name. I probably prefer the second option.

tilikang commented on include/functions.php:
```diff
@@ -0,0 +1,220 @@
+
+function getAllAuthorNames(){
+	$result = dbQuery("
+		SELECT DISTINCT authorOfPost
```
This approach definitely works for what you're doing right now, but eventually you might want a separate `users` table in your db. Each post would link to a UserId instead of just having the name of the author stored with the post. That way you can have additional info about the authors (bios, etc.) plus people could potentially log in to post comments and more.

tilikang commented on 17thblog/blogIndex.php:
```diff
@@ -0,0 +1,48 @@
+			echo "</div>
+		</div>
+</body>";
+
```
This page looks great. It looks like you've got a very strong grasp on how this stuff works.

tilikang commented on 17thblog/style.css:
```diff
@@ -58,19 +55,76 @@ a:hover{
 }
 
 /* This centers the words in the pic*/
 .centered {
```
CSS doesn't have namespacing which means you can only have a single class called `center`. For this reason, I'd suggest making CSS class names a bit more verbose. Something like `centeredPictureText` or something would make this more descriptive so you don't accidentally make a different `centered` element later on and have conflicts.

tilikang commented on include/functions.php:
```diff
@@ -0,0 +1,220 @@
+<?php
```
This file is starting to get a little bit long. I'd start breaking it up into things like `echo_html.php` and `blog.php` (or whatever delineation you want to make)

tilikang commented on include/functions.php:
```diff
@@ -0,0 +1,220 @@
+				<a href='blogIndex.php'>Blog |</a>
+				<a href='projects.php'>Projects</a>
+			</div>
+	</body>
```
You don't want to close the body here because there will be more text echoed later on, and the `</body>` should be at the very end. Instead, make a `echoFooter()` function or whatever and call that at the end of each page.

tilikang commented on include/functions.php:
```diff
@@ -0,0 +1,220 @@
+
+//Blog functions!!
+
+function getAllBlogPosts() {
```
This isn't necessary yet, but something to consider as a future project: what happens when you have 500 blog posts? Do you want to see them all on one page?

tilikang commented on include/functions.php:
```diff
@@ -0,0 +1,220 @@
+function insertBlogPost($title, $body, $dateCreated, $authorOfPost) {
+	$result = dbQuery("
+		INSERT INTO blogPost(title, body, dateCreated, authorOfPost)
+		VALUES('$title', '$body', '$dateCreated', '$authorOfPost')
```
This has a SQL injection vulnerability

samiklein011011 replied:
Got it. Should a function like this even exist? Because the only person who is writing blog posts is me, and I can just write them directly into SQL

tilikang replied:
Yes, it should. One of the future projects will be to make a web form that you can use to create/edit blog posts. The ultimate goal is for you to never need to access the db directly.

That's why a database is better than storing all this stuff in a PHP array. Right now, you have to manually to stuff either way, but a db has the option of being updated via a web form whereas PHP doesn't.

tilikang commented on include/functions.php:
```diff
@@ -0,0 +1,220 @@
+}
+
+//primary key will auto increment itself, don't need to add
+function insertBlogPost($title, $body, $dateCreated, $authorOfPost) {
```
Taking a `$dateCreated` field is fine, especially if you want the blogger to be able to change this date themselves. But in many cases, `$dateCreated` should just be generated inside this function rather than generating it outside the function and passing it in. You normally want to generate something like this as close to where it's being used as possible.

tilikang commented on include/functions.php:
```diff
@@ -0,0 +1,220 @@
+
+function getAllTagNames(){
+	$result = dbQuery("
+		SELECT DISTINCT tag
```
I can't see your schema here, but this doesn't seem like something that should be necessary. You should only ever store unique tag names in the db in the first place. Let's talk about this if you don't think that's how it's being stored right now.

tilikang commented on include/functions.php:
```diff
@@ -0,0 +1,220 @@
+}
+
+function getTagNamesWithThisNumber($oneTagNumber){
+	$result = dbQuery("
```
Ah, ok, I see why you have duplicate tag names. It's cool that you got this working, but there's a more elegant way to handle many-to-many relationships. Grab me sometime on Monday and I'll explain. Or if you'd rather figure it out on your own, read this: http://www.joinfu.com/2005/12/managing-many-to-many-relationships-in-mysql-part-1/

tilikang commented on include/functions.php:
```diff
@@ -0,0 +1,220 @@
+
+function printBlogPost($var){
+
+echo "<body class='bg'>";
```
There are times where you call this function inside a loop, meaning it might get called multiple times. You should only have one `<body>` tag ever on a page, so this probably shouldn't be inside this function.

tilikang commented on include/functions.php:
```diff
@@ -0,0 +1,220 @@
+
+
+
+function printBlogPost($var){
```
I'm not wild about using `$var` as the parameter. It doesn't say anything about what it is. Is it a `$blogPostId`? A `$blogPost`? Some other filter?

tilikang commented on include/functions.php:
```diff
@@ -0,0 +1,220 @@
+	$post=getOneBlogPost($var);
+
+	echo "<div class='textStyle'>";
+		echo "<strong>Author: </strong>";
```
This is fine, but a shorter way to write it would be:
```
echo "
   <div class='textStyle'>
      <strong>Author: </strong>
      $post[authorOfPost]
...and so on
```
I personally find all the echos to be a bit distracting and unnecessary.

samiklein011011 replied:
(P.S. I'm just putting a thumbs up now when I complete these changes, for my own reference. Let me know if it's annoyingly pinging you and I can stop!)

tilikang replied:
I don't get notifications when you leave the thumbs up, so it doesn't bother me at all.

tilikang commented on include/functions.php:
```diff
@@ -0,0 +1,220 @@
+		echo "<strong>Date Post Created: </strong>";
+		echo $post['dateCreated'];
+
+		$tagNames=getTagNamesWithThisNumber($post['blogPostID']);
```
Back to the topic of consistency with how you name things, I think I prefer the `getPostComments` name below. If you followed that convention, this would be `getPostTags`

tilikang commented on include/functions.php:
```diff
@@ -0,0 +1,220 @@
+		echo $post['body'];
+
+	if(getPostComments($post['blogPostID'])!= null) {
+		$comments = getPostComments($post['blogPostID']);
```
I'd move this to line 188, and then the if statement could check `$comments`. Right now you're running the exact same query twice in a row which means the database is getting hit twice. It's best to try to minimize the number of calls to the db unless there's a reason you need to call it multiple times.

tilikang commented on include/functions.php:
```diff
@@ -0,0 +1,220 @@
+
+
+function printPageName($wantToWrite){
+	echo "<body class='bg'>";
```
This is another case where the `<body>` tag will end up being used multiple times on one page. If you just need a wrapper around some html, use a `<div>`

tilikang reviewed (commented):
@samikleinLACRM - This is look great! It seems like the CRUD stuff is starting to make sense and you're seeing how you can layer more and more functionality into the site using the same tricks you already know.

Now that the concepts seem to be syncing in, I'd start focusing a bit more on code style, so I left a bunch of nitpicky comments about that. The only real functional thing that I think should be addressed is how the tags table is structured.

Commit:
6/5

Integrated some of the extra table, called post2tag. Did some other function re-structuring

Commit:
6/7 solid blog

autogenerates dates, can insert a blog post or a comment, organized include functions. a bunch of little stuff.

Commit:
6/8 end of the week

made a comments form, changed up the sort by tags and authors thing so can't see  multiple posts at once, but can see differnet links to singluar posts. organized functions, made a commendAdded page. commentsForm was made but now that stuff is in the singleBLogPost page. autogenerates dats, a bunch of other small things.

Commit:
6/11 Log In Created

Created a log in aspect for the website!

Commit:
6/11 night

Fixed one thing in the loggedIn.php file

Commit:
6/12 password visibility

made the password visibility toggleable, with javascript!!! Woo!

tilikang commented on 17thblog/singleBlogPageFrame.php:
```diff
@@ -0,0 +1,61 @@
+include('config/config.php');
+include('config/init.php');
+
+if($_REQUEST==null){
```
I'd probably specifically look for at least one field you need. Something like:

```
if(!@$_REQUEST['blogPostID']){
```

What you do works, but you could imagine having the request set, but not with the blogPostID

tilikang commented on 17thblog/singleBlogPageFrame.php:
```diff
@@ -0,0 +1,61 @@
+//echo's normal page
+$post=getOneBlogPost($_REQUEST['blogPostID']);
+echoHeader("Blog Page", $post['title']);
+echoBlogPost($_REQUEST['blogPostID']);
```
I'm not sure this really needs its own function. Functions are supposed to be for when you need to re-use code, and these non-include pages are for the one-time code that is only needed on this page. It's not necessarily wrong, but you're not getting anything out of having this as a function.

tilikang commented on 17thblog/singleBlogPageFrame.php:
```diff
@@ -0,0 +1,61 @@
+//comment form
+$Errors = array();
+
+if(isset($_REQUEST['commentForm'])){ //
```
As a general rule, I'd say that you should always process forms at the top of the page. IT's not strictly necessary if you don't redirect the page after submitting the form, but you pretty much always should redirect (even if you're redirecting back to this page) so that the user can't double-submit it by reloading the page.

tilikang commented on 17thblog/singleBlogPageFrame.php:
```diff
@@ -0,0 +1,61 @@
+			$_REQUEST['blogPostID'],
+			$_REQUEST['Author'],  //these have to match the form
+			$_REQUEST['Comment']
+		);
```
As mentioned above, I'd always redirect after saving something to the db, even if you redirect back to the page. If you want to skip down to the comment that was just saved, look into using <a name='Comment_123' />

Although once this is moved into js (if you choose to do that) it'll probably not matter anymore.

tilikang commented on include/logInFunctions.php:
```diff
@@ -0,0 +1,71 @@
+
+<script src="/include/jquery.js"></script>
```
script files shouldn't be directly in include files like this. For the most part, nothing should ever go in an include file except PHP functions. What this does is it echoes all this code to the page immediately when it runs, so it's not in the actual `<html>` tag that is supposed to wrap the whole page.

Instead, take all this js code and put it in your page header function.

tilikang commented on include/logInFunctions.php:
```diff
@@ -0,0 +1,71 @@
+	}
+}
+
+function createUser($username, $password){
```
This function probably doesn't need to exist. The code inside is good, but you've already got a separate function that does all the hard stuff (`insertUser`) so you could just move the header and exit lines into the main web page.

A good way to tell if a function serves a purpose is to look at the names. `createUser` and `insertUser` mean the same thing to most people, so if they're really doing different things, they should probably have more unique names.

tilikang commented on 17thblog/createAccount.php:
```diff
@@ -0,0 +1,49 @@
+if(isset($_REQUEST['createAccount'])){ //
+
+	validateFormField('Username');
+	validateFormField('Password');
```
you probably want an extra bit of validation here. What if the username is already in use? You should check for that and throw an error if it is.

tilikang commented on include/logInFunctions.php:
```diff
@@ -0,0 +1,71 @@
+//if this user exists w this pass in the database, then return true
+//if not, return false
+
+	$result=getUserRow($username);
```
This hopefully won't matter if you take my suggestion from above and validate that a username isn't already in use when someone signs up, but this function is an example of why this matters. Right now, if two users have the same username, you're just getting one of them here and checking against that password. So someone could potentially log in with the correct username/password combo and still not get logged in.

You asked if this is the correct way to verify a user. This definitely works for now, but later in the summer I'm going to go over some more advanced password security stuff. You'll want to make some adjustments after that.

tilikang commented on 17thblog/loggedIn.php:
```diff
@@ -0,0 +1,23 @@
+echo "<div class='textStyle form'>";
+
+//make a functino called verify user as lgoged in. so any page, u can call that function at top
+if(!isset($_SESSION['Username'])){
```
At some point you'll want to make this into a function called `verifyUserIsLoggedIn` or something like that so you can call it at the top of every page that requires a user to be logged in. You also might want to automatically redirect the user to a login page.

tilikang commented on 17thblog/logIn.php:
```diff
@@ -0,0 +1,76 @@
+
+			//set the session username = to that req and set the pass = to that
+			$_SESSION['Username'] = $_REQUEST['Username'];
+			$_SESSION['Password'] = $_REQUEST['Password'];
```
There shouldn't be any reason to store the password in the session. Passwords are normally something you want to interact with as little possible just so you don't run the risk of leaking it somewhere else.

Along these lines: You should probably use the UserId as the thing you use to identify a user. They'll log in with a username (humans don't like remembering random numbers for their logins) but internally, their UserId is their identifier. That should be the primary key in the db table, it should be the thing you store in the session, and it should be the parameter you pass to a function to get a user row most of the time.

For an example on why this matters: imagine a user changes their username. You'd have the username stored in the session, but it wouldn't be correct anymore, so they'd potentially be logged in as someone else (if someone else claims their old username). It's best to always use autoincrementing primary keys as the "pointer" to rows in your db.

tilikang commented on 17thblog/logIn.php:
```diff
@@ -0,0 +1,76 @@
+if(isset($_SESSION['Username']) && !empty($_SESSION['Username'])) {
+	echo "
+	<div class='textStyle form'>
+		**You are already logged in as $_SESSION[Username]!**
```
From a UI standpoint, I'd probably want to be redirected to my account homepage if I'm logged in. This makes it look like an error occurred or something which isn't really true.

tilikang commented on 17thblog/logIn.php:
```diff
@@ -0,0 +1,76 @@
+
+	if(sizeof($Errors) == 0){
+
+		if(verifyUser($_REQUEST['Username'], $_REQUEST['Password']) == true) {
```
The `== true` part isn't really necessary here. All `if` statements are evaluating if a statement is true, so this is like saying `if(true == true)` rather than just saying `if(true)`

tilikang reviewed (commented):
@samikleinLACRM - looking good! back to you.

Commit:
6/14

Fixed sessions so that it stores the userID, uses the password as little as possible, made it so that the admin can sucessfully edit a post!!! Ahh!
